pallets/file-system/src/utils.rs

```diff
@@ -0,0 +1,112 @@
+        match $optional {
+            Some(value) => value,
+            None => {
+                #[cfg(test)]
+                unreachable!($error_msg);
+
+                #[allow(unreachable_code)]
+                {
+                    Err($error_type)?
+                }
+            }
+        }
+    };
+}
+
+impl<T> Pallet<T>
+where
+    T: pallet::Config,
+{
+    pub fn do_request_storage(
+        location: FileLocation<T>,
+        fingerprint: Fingerprint<T>,
+        size: StorageUnit<T>,
+        user_multiaddr: MultiAddress<T>,
+    ) -> DispatchResult {
+        // TODO: Check user funds and lock them for the storage request.
+        // TODO: Check storage capacity of chosen MSP (when we support MSPs)
+        // TODO: Return error if the file is already stored and overwrite is false.
+        let current_block_number = <frame_system::Pallet<T>>::block_number();
+        let file_metadata = StorageRequestMetadata::<T> {
+            requested_at: current_block_number,
+            fingerprint,
+            size,
+            user_multiaddr,
+            bsps_volunteered: BoundedVec::default(),
+            bsps_confirmed: BoundedVec::default(),
+        };
+
+        // Check that storage request is not already registered.
+        ensure!(
+            !<StorageRequests<T>>::contains_key(&location),
+            Error::<T>::StorageRequestAlreadyRegistered
+        );
+
+        // Register storage request.
+        <StorageRequests<T>>::insert(&location, file_metadata);
+
+        // TODO: Maybe worth it to loop a few times until we find a free slot. (this should never fail and so a loop is required)
+        <StorageRequestExpirations<T>>::try_append(
+            current_block_number + T::StorageRequestTtl::get().into(),
+            &location,
+        )
+        .map_err(|_| Error::<T>::StorageRequestExpiredNoSlotAvailable)?;
+
+        Ok(())
+    }
+
+    pub fn do_bsp_volunteer(
+        who: T::AccountId,
+        location: FileLocation<T>,
+        _fingerprint: Fingerprint<T>,
+    ) -> DispatchResult {
+        // TODO: Verify BSP has enough storage capacity to store the file
+        // TODO: Check that sender is a registered storage provider
+
+        // Check that the storage request exists.
+        ensure!(
+            <StorageRequests<T>>::contains_key(&location),
+            Error::<T>::StorageRequestNotRegistered
+        );
+
+        // Get storage request metadata.
+        let mut file_metadata = expect_or_err!(
+            <StorageRequests<T>>::get(&location),
+            "Storage request should exist",
+            Error::<T>::StorageRequestNotRegistered
+        );
+
+        // Check if the BSP is not already registered for this storage request.
+        ensure!(
+            !file_metadata.bsps_volunteered.contains(&who),
+            Error::<T>::BspAlreadyConfirmed
+        );
+
+        // TODO: Check that the threshold value is high enough to qualify as BSP for the storage request.
```

Comment:
Just for clarity, the threshold should be higher than the BSP's calculated XOR value.